request_router.cc

```diff
@@ -1,103 +1,63 @@
-  if (resourcePath.substr(indexLastSlash).find(".") != std::string::npos) {
-    std::string resourceRoot = resourcePath.substr(0, indexLastSlash);
-    query.push_back("location " + resourceRoot);
-    std::cout << "HERE in if: " << "location " << resourceRoot << std::endl;
-  }
-  else {
-    query.push_back("location " + resourcePath);
-    std::cout << "HERE in else: " << "location " << resourcePath << std::endl;
   }
-  query.push_back("action");
 
+  // Build special 404 route, which is our fall-through error case
+  not_found_handler_ = RequestHandler::CreateByName("NotFoundHandler");
+  return true;
+}
 
-  std::cout << "~~begin query~~\n";
-  for (auto const& word : query) {
-    std::cout << word << ".";
-  }
-  std::cout << "\n~~end query~~\n";
-
-  std::string action;
-  server_config.propertyLookUp(query, action);
-  std::cout << "action: " << action << std::endl;
-
-  if (action == "echo") {
-    std::cout << "Dispatching echo handler\n";
-    EchoRequestHandler echoHandler;
-    echoHandler.handleRequest(server_config,
-                              request_buffer,
-                              request_buffer_size,
-                              response_buffer,
-                              response_buffer_size);
-    return true;
-  }
-  else if (action == "static_serve") {
-    std::cout << "Dispatching static file handler\n";
-    StaticRequestHandler staticHandler;
-    staticHandler.handleRequest(server_config,
-                                request_buffer,
-                                request_buffer_size,
-                                response_buffer,
-                                response_buffer_size);
-    return true;
-  }
-  else {
-    std::cout << "Dispatching 404 handler\n";
-    NotFoundRequestHandler notFoundHandler;
-    notFoundHandler.handleRequest(server_config,
-                                  request_buffer,
-                                  request_buffer_size,
-                                  response_buffer,
-                                  response_buffer_size);
-    return false;
+std::unique_ptr<RequestHandler> RequestRouter::routeRequest(const std::string& full_uri) const {
+  // Starting at the end of the string,
+  // try longest-prefix match up to latest forward-slash
+  //
+  // Example: for the full_uri '/echo/subaction/subdir/file.txt',
+  // we try '/echo/subaction/subdir', then '/echo/subaction',
+  // and finally '/echo'.
+  //
+  // Note the lack of trailing slashes
+  // See also: http://www.cplusplus.com/reference/string/string/find_first_of/
+  for (int search_index = full_uri.size() - 1; search_index >= 0; search_index--) {
+    size_t pos = full_uri.find_first_of('/', search_index);
+    if (pos == std::string::npos) {
+      continue;
+    }
+    else {
+      // Slice string from 0 to pos; substr takes start and length
+      // TODO: off-by-one?
+      std::string possible_prefix = full_uri.substr(0, full_uri.size() - pos);
```

Comment:
Not sure if I'm misunderstanding whats happening here but if the `possible_prefix` of something like `/static/file.txt`, I think the substring call would be something like `substr(0, pos)` to get `/static`